plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -1839,6 +1839,34 @@ private static boolean fieldExists(StructType structType, String fieldName)
         return false;
     }
 
+    @Override
+    public void setFieldType(ConnectorSession session, ConnectorTableHandle tableHandle, List<String> fieldPath, io.trino.spi.type.Type type)
+    {
+        Table icebergTable = catalog.loadTable(session, ((IcebergTableHandle) tableHandle).getSchemaTableName());
+        String parentPath = String.join(".", fieldPath.subList(0, fieldPath.size() - 1));
+        NestedField parent = icebergTable.schema().caseInsensitiveFindField(parentPath);
+
+        String caseSensitiveParentName = icebergTable.schema().findColumnName(parent.fieldId());
+        NestedField field = parent.type().asStructType().caseInsensitiveField(getLast(fieldPath));
+        if (!field.type().isPrimitiveType()) {
+            throw new TrinoException(NOT_SUPPORTED, "Iceberg doesn't support changing field type from non-primitive types");
+        }
+
+        String name = caseSensitiveParentName + "." + field.name();
+        Type icebergType = toIcebergTypeForNewColumn(type, new AtomicInteger());
+        if (!icebergType.isPrimitiveType()) {
+            throw new TrinoException(NOT_SUPPORTED, "Iceberg doesn't support changing field type to non-primitive types");
```

Comment:
for a table:

```sql
CREATE TABLE t (
    foo row(i integer, s varchar),
    nested_foo: row(foo row(i integer, s varchar))
);
```

we should probably allow
```sql
ALTER TABLE t SET DATA TYPE foo row(i bigint, s varchar); -- promote foo.i to bigint
ALTER TABLE t SET DATA TYPE nested_foo.foo row(i bigint, s varchar); -- promote nested_foo.foo.i to bigint
```

(not to be addressed in this PR)

Reply:
Added TODO comment.